src/any_llm/providers/anyllmplatform/utils.py

```diff
@@ -0,0 +1,261 @@
+        }
+    )
+
+    if response.status_code != 200:
+        logger.error(f"Error fetching provider key: {response.status_code}")
+        logger.error(response.json())
+        raise RuntimeError(response.text)
+
+    data = response.json()
+    logger.info("Provider key fetched")
+
+    return data
+
+
+def _decrypt_provider_key(encrypted_key: str, private_key) -> str:
+    """Decrypt the provider API key."""
+    logger.info("Decrypting provider API key...")
+
+    decrypted_key = _decrypt_data(encrypted_key, private_key)
+    logger.info("Decrypted successfully!")
+
+    return decrypted_key
+
+
+def get_provider_key(any_llm_key: str, provider: type[AnyLLM]) -> str:
+    """Get the provider key from Any LLM Platform.
+
+    The client first has to prove the ownership of the private key that decrypts the encrypted provider key,
+    without sharing it. To this end, the user asks Any LLM platform to create a new challenge, based on the public key,
+    and then the client attempts to solve it.
+
+    If the client solves the challenge correctly, Any LLM platform handles the encrypted provider key,
+    which can be decrypted using the private key (`ANY_LLM_KEY`).
+
+    Args:
+        any_llm_key: The Any LLM platform key, tied to a specific project.
+        provider: The name of the LLM provider.
+
+    Returns:
+        The provider key.
+    """
+    # Parse the `ANY_LLM_KEY`, load the private key and extract the public key from it.
+    _,_, private_key_base64 = _parse_any_llm_key(any_llm_key)
+    private_key = _load_private_key(private_key_base64)
+    public_key = _extract_public_key(private_key)
+
+    # Create and solve a new challenge to prove ownership of the private key without sharing it.
+    challenge_data = _create_challenge(public_key, ANY_LLM_PLATFORM_URL)
+    solved_challenge = _solve_challenge(challenge_data["encrypted_challenge"], private_key)
+
+    # Fetch and decrypt the provider key
+    provider_key_data = _fetch_provider_key(
+            provider=provider.PROVIDER_NAME,
+            public_key=public_key,
+            solved_challenge=solved_challenge,
+            any_api_url=ANY_LLM_PLATFORM_URL
+        )
+    return _decrypt_provider_key(
+            provider_key_data["encrypted_key"],
+            private_key
+        )
+
+
+async def post_completion_usage_event(any_llm_key: str, provider: str, completion: ChatCompletion) -> None:
+    """Posts completion usage events.
+
+    The client has to create and solve two challenges, one to prove ownership of the private key
+    that decrypts the encrypted provider key, without sharing it, and one to prove ownership of the project
+    it wants to post data to.
+
+    Args:
+        any_llm_key: The Any LLM platform key, tied to a specific project.
+        provider: The name of the LLM provider.
+        completion: The LLM response.
+    """
+    # Parse the `ANY_LLM_KEY`, load the private key and extract the public key from it.
```

Comment:
Do we need to run all these operations on each post_usage?

Can we do this one inside the `PlatformProvider` and consider the provider authenticated?